test/mpi/comm/cmsplit_type.c

```diff
@@ -55,6 +55,67 @@ int main(int argc, char *argv[])
         MPI_Comm_free(&comm);
     }
 
+    /* Check to see if MPI_COMM_TYPE_HW_GUIDED works correctly */
+    for (i = 0; split_topo[i]; i++) {
+        MPI_Info_create(&info);
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[i]);
+        MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+        if (comm != MPI_COMM_NULL) {
+            int newsize;
+            MPI_Comm_size(comm, &newsize);
+            if (newsize > size) {
+                printf("Expected comm size to be at most the node size\n");
+                errs++;
+            }
+            MPI_Comm_free(&comm);
+        }
+        MPI_Info_free(&info);
+    }
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: pass different info values, it must return
+     * MPI_COMM_NULL.*/
+    MPI_Info_create(&info);
+    if (rank % 2 == 0) {
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[0]);
+    } else {
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[1]);
+    }
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("Expected MPI_COMM_NULL due to different info values.\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    MPI_Info_free(&info);
```

Comment:
New test added

Reply:
As now this test covers several different info keys, maybe we want to also print at which case the error occurs. E.g., "Testing MPI_COMM_TYPE_HW_GUIDED: expected MPI_COMM_NULL due to different info values."

Reply:
An additional note: the standard says "All MPI processes must provide the same value for the info key "mpi_hw_resource_type." but it does not explicitly say if we should return `COMM_NULL` with `MPI_SUCCESS` or an MPI error. I guess the intention is an error. Otherwise such a case should be listed under "MPI_COMM_NULL is also returned in newcomm in the following cases:"